### ISSUE TITLE
unable to do Lab 5 level 4 equals method...

### ISSUE THREAD

**Issue by taniafoo** [nus-cs2030/2324-s2#343]
What does it mean by "logs are the same"? I was assuming to be the generic type wrapped inside the log for both to be the same, but I can't cast my type R as an instanceof to type T. 

Im not able to solve the test cases from the monad right identity onwards... send help plspls. 

Here is a code snippet:
public <R> boolean equals(Log<R> log) {
        R otherobject = log.getobject();
        String otherstring = log.getString();
        if (!this.t.equals(otherobject) || !this.log.equals(otherstring)) {
            return false;
        }
        /* if (!(otherobject instanceof T)) {
        return false;
        } */
        return true;
    }

**Comment by niclow236** [comment]
> What does it mean by "logs are the same"? I was assuming to be the generic type wrapped inside the log for both to be the same, but I can't cast my type R as an instanceof to type T.
> 
> Im not able to solve the test cases from the monad right identity onwards... send help plspls.
> 
> Here is a code snippet: public boolean equals(Log log) { R otherobject = log.getobject(); String otherstring = log.getString(); if (!this.t.equals(otherobject) || !this.log.equals(otherstring)) { return false; } /* if (!(otherobject instanceof T)) { return false; } */ return true; }

Maybe you can try using the templated equals method given in lecture 3 and adapt it for this lab.

jshell> class A {
   ...>     private final int x;
   ...>     A(int x) {
   ...>         this.x = x;
   ...>     }
   ...>     public boolean equals(Object obj) {
   ...>         if (this == obj) {
   ...>             return true;
   ...>         }
   ...>         if (obj instanceof A a) {
   ...>             return this.x == a.x;
   ...>         }
   ...>         return false;
   ...>     }
   ...> }
|  replaced class A

In this case, A would be Log, and if obj was instance of Log, the boolean returned would be whether the 2 Log Objects have the same value and logs.

**Comment by hanlinnnnn** [comment]
ur equals method shouldnt be taking in a log, but rather an object
also instanceof should only be used to check for instanceof Log and not on other classes

**Comment by gnimnix** [comment]
Take a look at [this ](https://stackoverflow.com/a/16383120/9141689)

Can also look through lecture 7 slides on how `equals` method is defined in `Maybe` class.

**Comment by leeyenkai** [comment]
I was in your exact position and the issue here (if you can't pass the test cases from the monad right identity onwards) is that your string values for your two logs may not exactly be the same even though it may look identical when debugging. Is your flatMap method adding an extra "\n" at the end of the string even when it shouldn't?

Check your `Log.<Integer>of(5).flatMap(x -> Log.<Integer>of(x)).toString()` and then compare it `Log.<Integer>of(5).toString()`. Are the strings the same?

Also, for other people facing this issue, remember to use `.equals()` when comparing the values and strings instead of `==`.

**Comment by MoeySeanJean** [comment]
Your equals method should overwrite original equals method from Object class. Thus, the parameter should be of class Object. You can refer to the equals class written by prof in the lecture, it uses instanceof and cast the Object to Log<?> class. After casting, you should be able to compare both the value and log using respective equals method.

**Comment by weiyangng** [comment]
u need to make sure your Log.of(5)).flatMap(x -> Log.of(x)) and Log.<Integer>of(5) is exactly the same. When I was stuck on this part my flatMap was adding extra blank lines at the end which resulted in the the equals returning false.

**Comment by taniafoo** [comment]
@weiyangng @leeyenkai youre totally right omg... thanks for pointing it out :) will redo my toString method

**Comment by ZHD1987E** [comment]
@taniafoo Did you manage to solve it? If so, great! You may close the issue now.
